scipy/sparse/_coo.py

```diff
@@ -722,16 +810,463 @@ def _matmul_multivector(self, other):
+        # Adjust negative axes for other
+        axes_other = [axis + other.ndim if axis < 0 else axis for axis in axes_other]
+
+        # Check for shape compatibility along specified axes
+        if any(self.shape[ax] != other.shape[bx]
+               for ax, bx in zip(axes_self, axes_other)):
+            raise ValueError("sizes of the corresponding axes must match")
+        
+        if isdense(other):
+            return self._dense_tensordot(other, axes_self, axes_other)
+        else:
+            return self._sparse_tensordot(other, axes_self, axes_other)
+
+
+    def _sparse_tensordot(self, other, axes_self, axes_other):
+        ndim_self = len(self.shape)
+        ndim_other = len(other.shape)
+
+        # Prepare the tensors for tensordot operation       
+        # Ravel non-reduced axes coordinates
+        self_non_red_coords = _ravel_non_reduced_axes(self.coords, self.shape,
+                                                      axes_self)
+        self_reduced_coords = np.ravel_multi_index(np.array(self.coords)[axes_self, :],
+                                                [self.shape[ax] for ax in axes_self])
+        other_non_red_coords = _ravel_non_reduced_axes(other.coords, other.shape,
+                                                       axes_other)
+        other_reduced_coords = np.ravel_multi_index(
+            np.array(other.coords)[axes_other, :],
+            tuple([other.shape[ax] for ax in axes_other]))
+        # Get the shape of the non-reduced axes
+        og_shape_self = tuple(self.shape[ax] for ax in range(ndim_self)
+                              if ax not in axes_self)
+        og_shape_other = tuple(other.shape[ax] for ax in range(ndim_other)
+                               if ax not in axes_other)
+        
+        # Create 2D coords arrays
+        ravel_coords_shape_self = (math.prod(og_shape_self),
+                                math.prod([self.shape[ax] for ax in axes_self]))
+        ravel_coords_shape_other = (math.prod([other.shape[ax] for ax in axes_other]),
+                                    math.prod(og_shape_other))
+
+        self_2d_coords = np.vstack((self_non_red_coords, self_reduced_coords))
+        other_2d_coords = np.vstack((other_reduced_coords, other_non_red_coords))
+
+        self_2d = coo_array((self.data, self_2d_coords), ravel_coords_shape_self)
+        other_2d = coo_array((other.data, other_2d_coords), ravel_coords_shape_other)
+
+        # Perform matrix multiplication (routed via 2-D CSR)
+        prod = (self_2d @ other_2d).tocoo()
+
+        # Combine the shapes of the non-contracted axes
+        combined_shape = og_shape_self + og_shape_other
+
+        # Unravel the 2D coordinates to get multi-dimensional coordinates
+        if og_shape_self:
+            unraveled_coords_self = np.array(np.unravel_index(prod.row, og_shape_self))
+        if og_shape_other:
+            unraveled_coords_other = np.array(np.unravel_index(prod.col,
```

Comment:
I've removed the `np.array` wrapping.
For the scalar case, I thought `coords` would just be an empty tuple, so using `isscalarlike(coords[0])` might give an IndexError. So I instead directly checked whether `coords == ()`. Is that fine?
Earlier, the code returned `prod.data[0]` but it doesn't seem to be the best way to return the scalar value, so I instead used `sum(prod.data)`. This would help deal with the case where we have no nnz elements so `data` is empty, and also in the case where we have multiple data values for the same coordinates.
Do these changes seem appropriate? - https://github.com/scipy/scipy/pull/21435/commits/5e8cd8a536857cdf0be8ce60124d107e16556cc7

Reply:
Hmmm... Yes!
The `sum(prod.data)` looks right to me. And checking `coords == ()` should do it too.

I'm having trouble wrapping my brain around the variables. 
What does `og` stand for in `og_shape_self`?  :)

I think a bunch of this could be written as a python "comprehension".
It looks like lines 1032-1044 can be combined and written as:
```python
    shapes = (og_shape_self, og_shape_other)
    coords = sum((np.unravel_index(c, s) for c, s in zip(prod.coords, shapes) if s), start=())
```
(I used `s` for shape and `c` for coord.)
Is it more clear/readable?  Does it work?
(that's often a matter of taste -- and I think the code is correct as it is, but I'm hoping this is easier to read.)
I'll let you decide.